File: 2.2.0-Elementary-functions.js
/* eslint func-call-spacing: ["error", "always"] */
/* eslint no-multi-spaces: ["error", { ignoreEOLComments: true }] */

/**
 * euclid ∷ (N x N) → N
 *
 * euclid (m, n) = euclid(n, m mod n), if (n > 0)
 *               = m                 , otherwise
 */

const { D, $ } = require ('./sanctuary') ()

//    euclid ∷ Integer → Integer → Integer
const euclid = D.def (
  'euclid',
  {},
  [$.Integer, $.Integer, $.Integer],
  m => n => (n > 0)
    ? euclid (n) (m % n)
    : m
)

console.log (
  euclid (14) (12),
  euclid (14) (11),
  euclid (558) (198),
)

File: 2.2.0_Elementary_functions.c

/**
 * euclid ∷ (N x N) → N
 *
 * euclid (m, n) = euclid(n, m mod n), if (n > 0)
 *               =                  m, otherwise
 */

/**
 * (* euclid : int -> int -> int *)
 * fun euclid m n = if n > 0
 *                     then euclid n (m mod n)
 *                     else m;
 */

#include <stdio.h>

int euclid (int m, int n) {
  if (n > 0) {
    return  euclid(n, m % n);
  } else {
    return m;
  }
}

int main (void) {
  printf("%d\n", euclid(14, 12));
  printf("%d\n", euclid(14, 11));
  printf("%d\n", euclid(558, 198));
  return 0;
}

